main/midi/midiPlayer.c

```diff
@@ -245,242 +239,272 @@ static adsrState_t voiceAdvanceAdsr(midiVoice_t* voice, voiceStates_t* states, u
+    // Decay State
+    /////////////////////////////////////////////
+adsrDecay:
+    decayTime = voice->envelope.decayTime + ((voice->envelope.decayTimeVel * (int)voice->velocity) >> 8);
 
-                MIDI_DBG("Skipping decay: no decay time");
-            }
-        }
-        else if (target == ADSR_SUSTAIN || (target == ADSR_ON && (states->decay & voiceBit)))
-        {
-            target = ADSR_ON;
-            states->decay &= ~voiceBit;
+    // Move from ATTACK to DECAY state
+    // This means that we move from the current volume (which is normally just the pressure volume)
+    // to the sustain volume
+    states->attack &= ~voiceBit;
 
-            // Ok so we're going from DECAY to SUSTAIN, sustain lasts forever
-            // BUT - we only go to sustain IF the note is still ON (via key on, hold, or sustenuto)
-            // Otherwise, we skip straight to release
-            if ((voiceBit & (states->on | states->held | states->sustenuto)) && sustainVol)
-            {
-                // The note is either being held on with a key, the hold pedal, or sustenuto pedal
-                // Go to sustain
-                states->sustain |= voiceBit;
-
-                // And either way, the note should be sustained indefinitely
-                // The key/pedal will be released eventually and we'll go to release then
-                voice->stateChangeTick = UINT32_MAX;
-
-                // Sustain lasts forever! As long as the note is on/held/sustenuto'd
-                // why is this the target volume? shouldn't it just be sustain volume?
-                // TODO check if this is better
-                voice->volRate  = 0;
-                voice->volAccel = 0;
-                // in theory we don't even need to set the current volume? since it should already have been set but idk
-                result = ADSR_SUSTAIN;
-
-                MIDI_DBG("Moving to sustain indefinitely");
-            }
-            else
-            {
-                // Go to release!
-                target = ADSR_RELEASE;
-                repeat = true;
+    if (decayTime)
+    {
+        states->decay |= voiceBit;
 
-                MIDI_DBG("Skipping sustain: no sustain volume or note already released");
-                MIDI_DBG("States: %" PRIx32 ", sustainVol: %" PRIu32,
-                         voiceBit & (states->on | states->held | states->sustenuto), sustainVol);
-            }
+        // calculate the volume rate of change
+        if (voice->curVol < sustainVol)
+        {
+            voice->volRate = (int)(sustainVol - voice->curVol) / decayTime;
         }
-        else if (target == ADSR_OFF || (target == ADSR_ON && (states->release & voiceBit)))
+        else if (voice->curVol > sustainVol)
         {
-            // Transition from RELEASE to OFF
-            states->release &= ~voiceBit;
-            states->on &= ~voiceBit;
-            states->held &= ~voiceBit;
-            states->sustain &= ~voiceBit;
```

Comment:
OK, so somewhere in here is an equivalent fix for 9c91a6264601710625ce29626d9953603f3d0c88, right? I'll double check that to be sure.

Reply:
Oh, I see `adsrOff:` now.